### PR TITLE
feat(cli): add support for running on macOS devices and visionOS simulators

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Why not.
- Better alignment with Xcode 15.

Screenshot 2024-04-24 at 12 13 54 PM

# How

- Use showdestinations to find the macos device ID and validate that the destination target is supported. This has a high coldboot time but it nets out fine because we'll invoke xcodebuild later in the process. 
- This adds a high degree of uncertainty since the functionality is new and I only have a small subset of devices to test against. Anything that isn't accounted for has a debug log so hopefully we have a bit of future proofing.
- We still sort iPhones to the top of the device list to preserve the happy path of `npx expo run:ios -d` + enter -> build on iPhone/iPad.
- We now show compat devices of visionOS and macOS even for iOS-only schemes. This is how Xcode 15 works.

# Test Plan

- Manually tested by bootstrapping a new project and running `nexpo run:ios -d` and selecting visionOS simulator, physical Apple Vision Pro (perhaps I can claim the device as a business expense now), iPhone OTA and connected, and the new macOS target.
- I added a parsing test for the xcodebuild results. These results are hard to parse correctly since phone names can contain any number of characters. I ran the matching through GPT a few times to sanity check if it was the safest approach. There are unit tests for the most aggressive behavior I was able to reproduce.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/run/ios/options/__tests__/resolveDevice-test.ts [hunk 1]
```diff
@@ -58,25 +62,50 @@ jest.mock('../../../../start/platforms/ios/AppleDeviceManager', () => ({
...
+          xcodeProject: { isWorkspace: true, name: '123 ' },
+        })
+      ).name
+    ).toEqual(`Evan's phone`);
 
     expect(promptDeviceAsync).toBeCalledWith([
       expect.anything(),
       expect.anything(),
-      expect.anything(),
+      // expect.anything(),
```
Comment (gabrieldonadel) [inline packages/@expo/cli/src/run/ios/options/__tests__/resolveDevice-test.ts:92]: probably want to remove this?
```suggestion 
```

**Review by brentvatne — APPROVED** [review]
